packages/strategy-saml/src/server/graphql/mutation/assertSaml.ts

```diff
@@ -0,0 +1,398 @@
+      Role.write(
+        tx,
+        {
+          ...role,
+          userIds: [...role.userIds, user.id],
+        },
+        {
+          recordId: v4(),
+          createdByAuthorizationId: newAuthorizationId,
+          createdAt: new Date(),
+        }
+      )
+    )
+  );
+
+  for (const result of roleResults) {
+    if (result.status === "rejected") {
+      throw new Error(result.reason);
+    }
+  }
+
+  await EmailCredential.write(
+    tx,
+    {
+      authorityId: emailAuthorityId,
+      authorityUserId: emailAddress,
+      details: {},
+      enabled: true,
+      id: v4(),
+      userId: user.id,
+    },
+    {
+      recordId: v4(),
+      createdAt: new Date(),
+      createdByAuthorizationId: newAuthorizationId,
+    }
+  );
+
+  return user;
+}
+
+async function createSamlCredentialForUser(
+  tx: PoolClient,
+  authority: SamlAuthority,
+  user: User,
+  newAuthorizationId: string,
+  samlId: string
+): Promise<SamlCredential> {
+  return await SamlCredential.write(
+    tx,
+    {
+      userId: user.id,
+      id: v4(),
+      enabled: true,
+      details: {},
+      authorityUserId: samlId,
+      authorityId: authority.id,
+    },
+    {
+      recordId: v4(),
+      createdAt: new Date(),
+      createdByAuthorizationId: newAuthorizationId,
+    }
+  );
+}
+
+export const assertSaml: GraphQLFieldConfig<
+  any,
+  Context,
+  {
+    authorityId: string;
+    samlResponse: string;
+  }
+> = {
+  type: GraphQLAuthorization,
+  description: "Create a new authorization.",
+  args: {
+    authorityId: {
+      type: new GraphQLNonNull(GraphQLID),
+    },
+    samlResponse: {
+      type: new GraphQLNonNull(GraphQLString),
+      description:
+        "The OAuth authorization code provided by the Saml exchange.",
+    },
+  },
+  async resolve(source, args, context): Promise<Authorization> {
+    const { executor, authorization: a, realm } = context;
+
+    if (a) {
+      throw new ForbiddenError("You area already authenticated.");
+    }
+
+    const strategies = executor.strategies;
+    const pool = executor.connection;
+    if (!(pool instanceof Pool)) {
+      throw new Error(
+        "INVARIANT: The executor connection is expected to be an instance of Pool."
+      );
+    }
+
+    const tx = await pool.connect();
+    try {
+      // Make sure this transaction is used for queries made by the executor.
+      const executor = new DataLoaderExecutor<Pool | PoolClient>(
+        tx,
+        strategies
+      );
+
+      await tx.query("BEGIN DEFERRABLE");
+
+      const authority = await SamlAuthority.read(executor, args.authorityId);
+
+      if (!(authority instanceof SamlAuthority)) {
+        throw new AuthenticationError(
+          "The authority uses a strategy other than SAML."
+        );
+      }
+
+      const options: GetAssertOptions = {
+        request_body: {
+          SAMLResponse: args.samlResponse,
+        },
+      };
+
+      const caller = async (): Promise<any> => {
```

Comment:
You may want to type out the expected response here instead of relying on `any`, which bypasses downstream property access/etc. Even just listing the properties you plan to use will help catch typos/etc.

Reply:
Fixed.